File: src/main/java/foo/changes/Change.java
package foo.changes;

import com.fasterxml.jackson.annotation.JsonTypeInfo;
import lombok.Getter;
import lombok.Setter;

import java.util.UUID;

@JsonTypeInfo(use = JsonTypeInfo.Id.NAME)
@Getter
@Setter
public abstract class Change {
    public final String changeId = newId();

    public abstract <T> T accept(ChangeVisitor<T> visitor);

    public static String newId() {
        return UUID.randomUUID().toString();
    }
}

File: src/main/java/ludwig/changes/Change.kt
package ludwig.changes

import com.fasterxml.jackson.annotation.JsonTypeInfo
import java.util.*

@JsonTypeInfo(use = JsonTypeInfo.Id.NAME)
abstract class Change {
    val changeId = newId()

    abstract fun <T> accept(visitor: ChangeVisitor<T>): T
}

fun newId(): String {
    return UUID.randomUUID().toString()
}
